code/modules/clothing/chameleon/_chameleon_action.dm

```diff
@@ -0,0 +1,201 @@
+	. = ..()
+	if(isnull(owner))
+		return
+
+	// Whenever a mob gains their first cham change action, they need to also gain the outfit action
+	if(locate(/datum/action/chameleon_outfit) in grant_to.actions)
+		return
+
+	var/datum/action/chameleon_outfit/outfit_action = new(owner)
+	outfit_action.Grant(owner)
+
+/datum/action/item_action/chameleon/change/Remove(mob/remove_from)
+	. = ..()
+	// Likewise when the mob loses the cham change action, if they have no others, they need to lose the outfit action
+	if(locate(/datum/action/item_action/chameleon/change) in remove_from.actions)
+		return
+
+	var/datum/action/chameleon_outfit/outfit_action = locate() in remove_from.actions
+	QDEL_NULL(outfit_action)
+
+/datum/action/item_action/chameleon/change/proc/initialize_disguises()
+	name = "Change [chameleon_name] Appearance"
+	build_all_button_icons()
+
+	LAZYINITLIST(chameleon_typecache)
+	LAZYINITLIST(chameleon_list)
+
+	chameleon_blacklist |= typecacheof(target.type)
+
+	if(!ispath(chameleon_type, /obj/item))
+		stack_trace("Non-item chameleon type defined on [type] ([chameleon_type])")
+		return
+
+	add_chameleon_items(chameleon_type)
+
+/datum/action/item_action/chameleon/change/proc/add_chameleon_items(type_to_add)
+
+	chameleon_typecache |= typecacheof(type_to_add)
+	for(var/obj/item/item_type as anything in chameleon_typecache)
+		if(chameleon_blacklist[item_type] || (initial(item_type.item_flags) & ABSTRACT) || !initial(item_type.icon_state))
+			continue
+		var/chameleon_item_name = "[initial(item_type.name)] ([initial(item_type.icon_state)])"
+		chameleon_list[chameleon_item_name] = item_type
+
+
+/datum/action/item_action/chameleon/change/proc/select_look(mob/user)
+	var/picked_name = tgui_input_list(user, "Select [chameleon_name] to change into", "Chameleon Settings", sort_list(chameleon_list, GLOBAL_PROC_REF(cmp_typepaths_asc)))
+	if(isnull(picked_name) || isnull(chameleon_list[picked_name]) || QDELETED(src) || QDELETED(user) || QDELETED(owner) || !IsAvailable(feedback = TRUE))
+		return
+	var/obj/item/picked_item = chameleon_list[picked_name]
+	update_look(picked_item)
+
+/datum/action/item_action/chameleon/change/proc/random_look()
+	var/picked_name = pick(chameleon_list)
+	update_look(chameleon_list[picked_name])
+
+/datum/action/item_action/chameleon/change/proc/update_look(obj/item/picked_item)
+	var/obj/item/chameleon_item = target
+
+	update_item(picked_item)
+	build_all_button_icons()
+	active_type = picked_item
+
+	if(ismob(chameleon_item.loc))
+		var/mob/wearer = chameleon_item.loc
+		wearer.update_clothing(chameleon_item.slot_flags | ITEM_SLOT_HANDS)
+
+/datum/action/item_action/chameleon/change/proc/update_item(obj/item/picked_item)
+	PROTECTED_PROC(TRUE) // Call update_look, not this!
+
+	var/atom/atom_target = target
+	atom_target.name = initial(picked_item.name)
+	atom_target.desc = initial(picked_item.desc)
+	atom_target.icon_state = initial(picked_item.icon_state)
+	if(isitem(atom_target))
```

Comment:
this is such a slog to read would you kindly add some newlines and break it into chunks